libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -323,23 +546,242 @@ static MQTTStatus_t addRecord( MQTTPubAckInfo_t * records,
     return status;
 }
 
+/*-----------------------------------------------------------*/
+
 static void updateRecord( MQTTPubAckInfo_t * records,
                           size_t recordIndex,
                           MQTTPublishState_t newState,
                           bool shouldDelete )
 {
+    assert( records != NULL );
+
     if( shouldDelete == true )
     {
+        /* Mark the record as invalid. */
         records[ recordIndex ].packetId = MQTT_PACKET_ID_INVALID;
-        records[ recordIndex ].qos = MQTTQoS0;
-        records[ recordIndex ].publishState = MQTTStateNull;
     }
     else
     {
         records[ recordIndex ].publishState = newState;
     }
 }
 
+/*-----------------------------------------------------------*/
+
+static uint16_t stateSelect( const MQTTContext_t * pMqttContext,
+                             uint16_t searchStates,
+                             MQTTStateCursor_t * pCursor )
+{
+    uint16_t packetId = MQTT_PACKET_ID_INVALID;
+    uint16_t outgoingStates = 0U;
+    const MQTTPubAckInfo_t * records = NULL;
+    bool stateCheck = false;
+
+    assert( pMqttContext != NULL );
+    assert( searchStates != 0U );
+    assert( pCursor != NULL );
+
+    /* Create a bit map with all the outgoing publish states. */
+    UINT16_SET_BIT( outgoingStates, MQTTPublishSend );
+    UINT16_SET_BIT( outgoingStates, MQTTPubAckPending );
+    UINT16_SET_BIT( outgoingStates, MQTTPubRecPending );
+    UINT16_SET_BIT( outgoingStates, MQTTPubRelSend );
+    UINT16_SET_BIT( outgoingStates, MQTTPubCompPending );
+
+    /* Only outgoing publish records need to be searched. */
```

Comment:
Should we also assert that only outgoing states are present in searchStates? (i.e. `(~outgoingStates & searchStates) == 0` )

Reply:
Yes. That will be a good addition.